### ISSUE TITLE
grpc_auth_json_key_destruct not found

### ISSUE THREAD

**Issue by cogmeta** [googlesamples/assistant-sdk-cpp#15]
src/service_account_util.o: In function `GetServiceAccountCredentialsOrNull(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)::{lambda()#1}::operator()() const':
service_account_util.cc:(.text+0xb0): undefined reference to `grpc_auth_json_key_destruct'
src/service_account_util.o: In function `GetServiceAccountCredentialsOrNull(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)::{lambda()#3}::operator()() const':
service_account_util.cc:(.text+0x108): undefined reference to `grpc_json_destroy'
src/service_account_util.o: In function `GetServiceAccountCredentialsOrNull(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)':
service_account_util.cc:(.text+0x144): undefined reference to `grpc_auth_json_key_create_from_string'
service_account_util.cc:(.text+0x188): undefined reference to `grpc_auth_json_key_is_valid'
service_account_util.cc:(.text+0x21c): undefined reference to `grpc_jwt_encode_and_sign'
service_account_util.cc:(.text+0x4b0): undefined reference to `grpc_json_parse_string_with_len'
service_account_util.cc:(.text+0x554): undefined reference to `grpc_json_get_string_property'
collect2: error: ld returned 1 exit status
Makefile:59: recipe for target 'run_assistant' failed

**Comment by Fleker** [comment]
Did you clone & compile gRPC?

**Comment by cogmeta** [comment]
Yes. I see the functions are defined in files located in 'src/core/lib/security/credentials/jwt' dir but i cant find them in any of the lib files in 'libs/opt'.

**Comment by cogmeta** [comment]
Did a clean install on another machine but ended up with same error. Is it just me?

**Comment by Fleker** [comment]
I've done many of these builds in isolated environments and never encountered this build error. Is it an Ubuntu/Linux-based machine?

**Comment by cogmeta** [comment]
Something is clearly broken. I was initially trying on Raspberian. But I just gave a try on Ubuntu 16.04. Everything goes smooth in grpc, googleapis untll the same error again in assistant-sdk-cpp. You need to try it now. Maybe something has changed since last time you built it.

**Comment by cogmeta** [comment]
any minimum version requirement on g++  or any other?

**Comment by cogmeta** [comment]
Meanwhile i am using fixes suggested in #6 . I was just wondering why that has not been fixed yet in Makefile.

**Comment by Fleker** [comment]
I'll push out some changes to the sample on Monday. The changed files have definitely been building.

I don't know the minimum g++ version that is required offhand.

**Comment by cogmeta** [comment]
Thanks. Regardless of the platform, I think this is more likely linker issue. The symbols are defined in the libraries but somehow the linker is not able to find it. I need brush up little on the linking order. I think the issue is with order of libraries laid out in  compilation command. Let me play around around with the command and see if that fixes it.

**Comment by cogmeta** [comment]
> nm -D  --extern-only --defined-only libgrpc.so | grep grpc_json_destroy
0000000000062900 T _Z17grpc_json_destroyP9grpc_json

No idea why the symbols is defined this way in the library

**Comment by Fleker** [comment]
I've updated the build info. Can you take a look at the update and see if there's any changes that may resolve your issue?

**Comment by cogmeta** [comment]
Ok, it finally compiles (after getting rid of  ./src/service_account_util.o which no longer exists :))

I can not test it though since I do not have mic on this device. Will try again from scratch on another device and test it. Thanks.

**Comment by cogmeta** [comment]
The first run was successful then i have this problem

g++ src/embedded_assistant.pb.o src/embedded_assistant.grpc.pb.o googleapis.ar src/audio_input_alsa.o src/audio_output_alsa.o src\
/audio_input_file.o src/json_util.o src/run_assistant.o  -lm -L/usr/local/lib `pkg-config --libs grpc++ grpc` -lgrpc_cronet -Wl,-\
-no-as-needed -lgrpc++_reflection -Wl,--as-needed -lprotobuf -lpthread -ldl -lcurl `pkg-config --libs alsa` -o run_assistant
src/embedded_assistant.pb.o: In function `protobuf_embedded_5fassistant_2eproto::protobuf_AssignDescriptors()':
embedded_assistant.pb.cc:(.text+0x6f4): undefined reference to `google::protobuf::internal::AssignDescriptors(std::string const&,\
 google::protobuf::internal::MigrationSchema const*, google::protobuf::Message const* const*, unsigned int const*, google::protob\
uf::MessageFactory*, google::protobuf::Metadata*, google::protobuf::EnumDescriptor const**, google::protobuf::ServiceDescriptor c\
onst**)'
src/embedded_assistant.pb.o: In function `protobuf_embedded_5fassistant_2eproto::AddDescriptorsImpl()':
embedded_assistant.pb.cc:(.text+0x7c8): undefined reference to `google::protobuf::MessageFactory::InternalRegisterGeneratedFile(c\
har const*, void (*)(std::string const&))'
src/embedded_assistant.pb.o: In function `google::assistant::embedded::v1alpha1::ConverseState::MergePartialFromCodedStream(googl\
e::protobuf::io::CodedInputStream*)':
embedded_assistant.pb.cc:(.text+0x5160): undefined reference to `google::protobuf::internal::WireFormatLite::ReadBytes(google::pr\
otobuf::io::CodedInputStream*, std::string*)'
src/embedded_assistant.pb.o: In function `google::assistant::embedded::v1alpha1::ConverseState::SerializeWithCachedSizes(google::\
protobuf::io::CodedOutputStream*) const':
embedded_assistant.pb.cc:(.text+0x5378): undefined reference to `google::protobuf::internal::WireFormatLite::WriteBytesMaybeAlias\
ed(int, std::string const&, google::protobuf::io::CodedOutputStream*)'
src/embedded_assistant.pb.o: In function `google::assistant::embedded::v1alpha1::AudioOut::MergePartialFromCodedStream(google::pr\
otobuf::io::CodedInputStream*)':
embedded_assistant.pb.cc:(.text+0x65fc): undefined reference to `google::protobuf::internal::WireFormatLite::ReadBytes(google::pr\
otobuf::io::CodedInputStream*, std::string*)'
src/embedded_assistant.pb.o: In function `google::assistant::embedded::v1alpha1::AudioOut::SerializeWithCachedSizes(google::proto\
buf::io::CodedOutputStream*) const':

**Comment by cogmeta** [comment]
Nope. Did not touch grpc or protobuf at all.

**Comment by Fleker** [comment]
It seems like something has changed in the `embedded_assistant.cc` file. Can you check it?

**Comment by cogmeta** [comment]
My apologies, this is probably due to changes to Makefile  we made. But, it is strange to see the error even when the protobuf and grpc libs are included.

**Comment by cogmeta** [comment]
There is an error when linking pb.cc (pb.o) files generated from .proto files. Even when the protobuf library mentioned with correct -L/usr/local/lib, there is an 'undefined reference' error. Dont know what is going on.

**Comment by Fleker** [comment]
Yes that is pretty odd. I think the proto output is different than expected, where it is omitting certain methods.
